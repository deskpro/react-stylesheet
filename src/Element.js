/**
 * @flow
 */

const React = require('react');

import type {ClassName} from './CSSStyleRepr';
import CSSPseudoClassSet from './CSSPseudoClassSet';
import CSSPropertySet from './CSSPropertySet';
import * as CSS from './CSS';
import * as Stylesheet from './Stylesheet';
import defaultBoxStyle from './DefaultBoxStyle';

/**
 * <Element /> API
 */

type ElementProps<P: {}> = {
  Component: string | React.ComponentType<P>,
  className?: ClassName,
};

export class Element<
  P: {},
  Props: CSS.CSSStylesheet & P & ElementProps<P>,
> extends React.Component<Props> {
  context: Stylesheet.StylesheetContext;

  static defaultProps = {
    Component: 'div',
  };

  dispose: null | (() => void);
  stylesheet: Stylesheet.Stylesheet;
  restProps: P;

  constructor(props: Props) {
    super(props);
    const [restProps, stylesheet] = this.createStylesheetFromProps(props);
    this.restProps = restProps;
    this.stylesheet = stylesheet;
    this.dispose = null;
  }

  render() {
    const {Component, className: extraClassName} = this.props;
    const className = Stylesheet.toClassName(
      this.stylesheet,
      {},
      {rightToLeft: this.context.rightToLeft, className: extraClassName},
    );
    return <Component {...this.restProps} className={className} />;
  }

  componentWillMount() {
    this.dispose = Stylesheet.injectDisposableStylesheet(this.stylesheet);
  }

  componentWillReceiveProps(nextProps: Props) {
    const [restProps, stylesheet] = this.createStylesheetFromProps(nextProps);
    if (this.stylesheet == null || stylesheet.id !== this.stylesheet.id) {
      if (this.dispose != null) {
        this.dispose();
      }
      this.stylesheet = stylesheet;
      this.dispose = Stylesheet.injectDisposableStylesheet(this.stylesheet);
    }
    this.restProps = restProps;
  }

  componentWillUnmount() {
    if (this.dispose != null) {
      this.dispose();
    }
  }

  createStylesheetFromProps(props: any): [P, Stylesheet.Stylesheet] {
    const style = {...props.style};
    const restProps: any = {};
    const spec = {base: {}};
    let needStyle = false;
    for (const key in props) {
      if (CSSPropertySet[key]) {
        style[key] = props[key];
        needStyle = true;
      } else if (CSSPseudoClassSet[key]) {
        spec.base[key] = props[key];
      } else if (key !== 'Component') {
        restProps[key] = props[key];
      }
    }
    if (needStyle) {
      restProps.style = style;
    }
    return [restProps, Stylesheet.createStylesheet(spec)];
  }
}

const VBoxStylesheet = Stylesheet.createStylesheet({
  displayName: 'VBoxBase',
  base: {...defaultBoxStyle, flexDirection: 'column'},
});

const HBoxStylesheet = Stylesheet.createStylesheet({
  displayName: 'HBoxBase',
  base: {...defaultBoxStyle, flexDirection: 'row'},
});

Stylesheet.injectStylesheet(VBoxStylesheet);
Stylesheet.injectStylesheet(HBoxStylesheet);

export class VBox<P: {}> extends Element<P, *> {
  static defaultProps = {
    Component: 'div',
    className: Stylesheet.toClassName(VBoxStylesheet, {}),
  };
}

export class HBox<P: {}> extends Element<P, *> {
  static defaultProps = {
    Component: 'div',
    className: Stylesheet.toClassName(HBoxStylesheet, {}),
  };
}
