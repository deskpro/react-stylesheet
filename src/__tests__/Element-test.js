/**
 * @flow
 */

const React = require('react');
const renderer = require('react-test-renderer');

import {Element} from '../Element';
import * as TestUtil from '../TestUtil';

// $FlowFixMe: ...
expect.addSnapshotSerializer(TestUtil.snapshotSerializer);

describe('<Element /> inserts styles', function() {
  test('<Element color={...} />', function() {
    const component = renderer.create(<Element color="red" />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('<Element hover={...} />', function() {
    const component = renderer.create(<Element hover={{color: 'red'}} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('<Element color={...} hover={...} />', function() {
    const component = renderer.create(<Element color="black" hover={{color: 'red'}} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('<Element focus={...} hover={...} />', function() {
    const component = renderer.create(
      <Element focus={{color: 'black'}} hover={{color: 'red'}} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('<Element className={...} />', function() {
    const component = renderer.create(<Element className="bootstrap" />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('<Element color="red" className={...} />', function() {
    const component = renderer.create(<Element color="red" className="bootstrap" />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
