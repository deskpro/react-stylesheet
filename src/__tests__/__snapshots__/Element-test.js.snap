// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Element /> inserts styles <Element className={...} /> 1`] = `
<div
  className={
    CSSClassJoin [
      "bootstrap"
    ]
  }
/>
`;

exports[`<Element /> inserts styles <Element color="red" className={...} /> 1`] = `
<div
  className={
    CSSClassJoin [
      "bootstrap"
    ]
  }
  style={
    Object {
      "color": "red",
    }
  }
/>
`;

exports[`<Element /> inserts styles <Element color={...} /> 1`] = `
<div
  className={null}
  style={
    Object {
      "color": "red",
    }
  }
/>
`;

exports[`<Element /> inserts styles <Element color={...} hover={...} /> 1`] = `
<div
  className={
    CSSClassJoin [
      CSSClass base:hover {
        color: red !important
      }
    ]
  }
  style={
    Object {
      "color": "black",
    }
  }
/>
`;

exports[`<Element /> inserts styles <Element focus={...} hover={...} /> 1`] = `
<div
  className={
    CSSClassJoin [
      CSSClass base:focus {
        color: black !important
      }
      CSSClass base:hover {
        color: red !important
      }
    ]
  }
/>
`;

exports[`<Element /> inserts styles <Element hover={...} /> 1`] = `
<div
  className={
    CSSClassJoin [
      CSSClass base:hover {
        color: red !important
      }
    ]
  }
/>
`;
